New feature CORE-4933 : Add better transaction control to isql

hvlad opened the pull request:
Add new SET command at ISQL: SET KEEP_TRAN_PARAMS:
- when ON it keeps text of following successful SET TRANSACTION statement and
new DML transactions is started using the same SQL (instead of defaul CONCURRENCY WAIT mode)
- when OFF, isql start new DML transaction as usual.
Name KEEP_TRAN_PARAMS could be cut up to the KEEP_TRAN.

Example:

SQL> SET;
...
**Keep transaction params: OFF**
SQL> SET KEEP_TRAN;
SQL> SET;
...
**Keep transaction params: ON
    SET TRANSACTION**
SQL>commit;
SQL>SET TRANSACTION READ COMMITTED WAIT;
SQL>SET;
...
**Keep transaction params: ON
    SET TRANSACTION READ COMMITTED WAIT**
SQL> SELECT RDB$GET_CONTEXT('SYSTEM', 'ISOLATION_LEVEL') FROM RDB$DATABASE;

RDB$GET_CONTEXT

=============================================================
READ COMMITTED
SQL> commit;
SQL> SELECT RDB$GET_CONTEXT('SYSTEM', 'ISOLATION_LEVEL') FROM RDB$DATABASE;

RDB$GET_CONTEXT

=============================================================
READ COMMITTED

SQL> SET KEEP_TRAN OFF;
SQL> SELECT RDB$GET_CONTEXT('SYSTEM', 'ISOLATION_LEVEL') FROM RDB$DATABASE;

RDB$GET_CONTEXT

=============================================================
READ COMMITTED

SQL> commit;
SQL> SELECT RDB$GET_CONTEXT('SYSTEM', 'ISOLATION_LEVEL') FROM RDB$DATABASE;

RDB$GET_CONTEXT

=============================================================
SNAPSHOT

SQL> SET;
...
Keep transaction params: OFF
SQL>

hvlad committed:
New feature CORE-4933 : Add better transaction control to isql

aafemt commented:
Isn't it a little too complicated? May be "SET TRANSACTION [AS PREVIOUS]" could be better...?

hvlad commented:
> 
> 
> Isn't it a little too complicated? May be "SET TRANSACTION [AS PREVIOUS]" could be better...?

SET TRANSACTION is handled by the engine. I don't want to add further confusion creating "isql" dialect for this statement.

AlexPeshkoff commented:
May be better 
SET DEFAULT TRANSACTION ...
followed by same set of parameters as in SET TRANSACTION but instead starting new transaction this statements changes parameters for default transactions?

This appears much more straightforward compared with your suggestion and helps avoid obvious errors when SET TRANSACTION is erroneously used when KEEP_TRAN is ON (default transactions start to have other parameters in unexpected by user moment). On the other hand this does not create ISQL-specific dialect of SET TRANSACTION cause we use another (SET DEFAULT) statement for it.

hvlad commented:
> May be better
> SET DEFAULT TRANSACTION ...
> followed by same set of parameters as in SET TRANSACTION but instead starting new transaction this statements changes parameters for default transactions?

I've considered this. The problem is that isql have no idea of correct syntax of SET TRANSACTION and can't verify correctness of proposed SET DEFAULT TRANSACTION. I.e. it could keep wrong statement text and error will be raised at some other moment - when new transaction will be started. Example:

SET DEFAULT TRANSACTION READ COMITTED; -- OK, but note: M is missed in CO**M**MITTED
COMMIT; -- OK
SELECT ... -- error about wrong transaction parameters

SET KEEP_TRAN can't save wrong sql text as it get it only after successful SET TRANSACTION execution.

> This appears much more straightforward compared with your suggestion and helps avoid obvious errors when SET TRANSACTION is erroneously used when KEEP_TRAN is ON (default transactions start to have other parameters in unexpected by user moment). 

So far, I consider SET KEEP_TRAN less error prone than SET DEFAULT TRAN.

aafemt commented:
> The problem is that isql have no idea of correct syntax of SET TRANSACTION and can't 
> verify correctness of proposed SET DEFAULT TRANSACTION.

This statement can implicitly start a new transaction verifying the parameters by server.

hvlad commented:
> 
> The problem is that isql have no idea of correct syntax of SET TRANSACTION and can't verify correctness of proposed SET DEFAULT TRANSACTION.
> This statement can implicitly start a new transaction verifying the parameters by server.
 
And we will explain how SET DEFAULT TRANSACTION is different from SET TRANSACTION, when user should use one statement and when another (and why),  why SET DEFAULT TRANSACTION "not works" in user applicaitons and so on and so on.

It creates much more problems than it solves.

aafemt commented:
> And we will explain how SET DEFAULT TRANSACTION is different from SET TRANSACTION, when 
> user should use one statement and when another (and why), why SET DEFAULT TRANSACTION "not 
> works" in user applicaitons and so on and so on.

Yes. That's what documentation is supposed to be for.

hvlad commented:
@aafemt are you arguing just for the sake of the process?
I clearly explained why SET DEFAULT TRAN is not good. Lets not create problems just to fight with it later.

AlexPeshkoff commented:
On 2020-02-17 13:45, Vlad Khorsun wrote:
>
>     The problem is that isql have no idea of correct syntax of SET
>     TRANSACTION and can't verify correctness of proposed SET DEFAULT
>     TRANSACTION.
>     This statement can implicitly start a new transaction verifying
>     the parameters by server.
>
> And we will explain how SET DEFAULT TRANSACTION is different from SET 
> TRANSACTION, when user should use one statement and when another (and 
> why), why SET DEFAULT TRANSACTION "not works" in user applicaitons and 
> so on and so on.
>
> It creates much more problems than it solves.
>

We may avoid explaining that much details to users if successfully 
started transaction is rolled back at once.

hvlad commented:
> We may avoid explaining that much details to users if successfully started transaction is rolled back at once.

For what such complexity and overhead ?

AlexPeshkoff commented:
On 2020-02-17 13:53, Vlad Khorsun wrote:
>
>     We may avoid explaining that much details to users if successfully
>     started transaction is rolled back at once.
>
> For what such complexity and overhead ?
>

I do not find mixing in single statement 2 different functions (start 
transaction and set parameters for future statements), specially when 
performing or not one of them depends upon stored state set by another 
statement. Onmy mind _this_ is unwanted complexity. But if you are sure 
that it's ok I do not plan to argue any more - may be I miss something.

hvlad commented:
@AlexPeshkoff 
What kind of complexity you see with SET KEEP_TRAN ? 
It will not affect next transaction params until user explicitly set new params with SET TRANSACTION.
I.e. there is no "stored state" until user explicitly "store" it.

Compare:

```
select current_transaction, iso_level ...
1000, concurrency
...
set keep_tran on;

select current_transaction, iso_level ...
1000, concurrency

... many many statements in the same tran...

commit;

// new tran params not changed, as expected
select current_transaction, iso_level ...
1010, concurrency
commit;

set transaction read committed;

// new tran params is changed, as expected
select current_transaction, iso_level ...
1020, read committed
```

with 

```
select current_transaction, iso_level ...
1000, concurrency
...
set default transaction read committed;

select current_transaction, iso_level ...
1000, concurrency

... many many statements in the same tran...

commit;

// new tran params is changed, is it expected ?
select current_transaction, iso_level ...
1010, read committed
```

aafemt commented:
> What kind of complexity you see with SET KEEP_TRAN ?

Why not turn it on permanently? In this case you won't need this new command at all.

New transaction after commit automatically start with parameters of previous one and if 
someone want to change them, he issue SET TRANSACTION statement. What's wrong with it?

hvlad commented:
>> What kind of complexity you see with SET KEEP_TRAN ?

> Why not turn it on permanently? In this case you won't need this new command at all. New transaction after commit automatically start with parameters of previous one and if someone want to change them, he issue SET TRANSACTION statement. What's wrong with it?

It breaks backward compatibility, unfortunately.

aafemt commented:
> It breaks backward compatibility, unfortunately.

"Break backward compatibility" means that reset of transaction parameters to default ones 
after commit/rollback is documented somewhere as a feature. Change CORE-4933 from "new 
feature: preserve transaction parameters" to "bug report: transaction parameters 
unexpectedly reset after commit" and there is no problem with backward compatibility any more.

hvlad commented:
@aafemt 
You perfectly know what is backward compatibility.
And not, it can't be fixed changing ticket in tracker.

aafemt commented:
> You perfectly know what is backward compatibility.

Exactly because of this I'm sure that it can be fixed by changing of the ticket.

I cannot say for everybody but I personally always expected transaction parameters set 
with "set transaction" to be persistent until the next explicit "set transaction" 
statement. That's why I'm sure that resetting them to default after commit is a bug that 
must be fixed.

hvlad commented:
Change well known for tens of years behavior (even if not explicitly documented) - is not good idea. Especially in such sensitive area as transaction isolation mode.

I could agree with introducing SET KEEP_TRANS in v4, and making it ON by default in v5, for example.
But your offer breaks backward compatibility.

aafemt commented:
I would suggest on contrary: turn it on in v4 and for the sake of backward compatibility allow to turn it off with the command until v5 because number of people who are aware of this behavior is most likely much smaller than those who aren't.

hvlad commented:
If team accept such a change - I won't resist

sim1984 commented:
In this case, I agree with @aafemt. Saving transaction parameters until explicitly executing SET TRANSACTION is what the user expects as the default behavior. When I discovered that this was not so, I was very surprised. For backward compatibility, you can give the ability to disable it.

hvlad commented:
Backward compatibility is when things works in the exactly same way. What @aafemt offers is not backward compatible, despite of users expectation, despite of what you know or do not know. Please, understand it.

hvlad committed:
Make KEEP_TRANS_PARAMS = ON by default.
Don't reset saved transaction parameters if run "SET KEEP_TRAN ON" when KEEP_TRANS_PARAMS was already set to ON.

hvlad commented:
After discussion I decided to 
- leave SET KEEP_TRAN_PARAMS syntax in isql
- make default value ON in master (fb4)
- make default value OFF in B3_0_Release (fb3)

artyom-smirnov commented:
I've noticed  test bugs.core_5533 hanging after merging this branch. In debug it crashing on assert. Git bisect tells it broken in 1c844b218da29855e8cdbb69b062b849726d837b. May be some regression surfaced after these changes?
```
Thread 5 "firebird" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff13de700 (LWP 30457)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7734899 in __GI_abort () at abort.c:79
#2  0x00007ffff52bf6a5 in fb_assert_impl (msg=0x7ffff5b23248 "tra_interface == NULL || tra_interface == jt", 
    file=0x7ffff5b22a90 "/workspace/rdb/firebird_master/src/jrd/tra.cpp", line=3695, do_abort=true) at /workspace/rdb/firebird_master/src/include/../common/gdsassert.h:48
#3  0x00007ffff556111c in Jrd::jrd_tra::setInterface (this=0x7fffe2951ca0, jt=0x7ffff505b3b0) at /workspace/rdb/firebird_master/src/jrd/tra.cpp:3695
#4  0x00007ffff54b2ef6 in Jrd::JAttachment::execute (this=0x7ffff507ebe0, user_status=0x7ffff13dd6e0, apiTra=0x0, length=15, 
    string=0x7ffff7700d10 "SET TRANSACTIONrg_tx_start activeinct_count_0 from tlogrt position 0\n    as\n    begin\n      execute statement ('insert into tlog(msg) values(?)') ('Tx start');\n    end\n    \314\314\314\314", '\335' <repeats 15 times>, "݀\366\205UUU", dialect=3, inMetadata=0x0, inBuffer=0x0, outMetadata=0x0, outBuffer=0x0)
    at /workspace/rdb/firebird_master/src/jrd/jrd.cpp:5071
#5  0x00007ffff54d186d in Firebird::IAttachmentBaseImpl<Jrd::JAttachment, Firebird::CheckStatusWrapper, Firebird::IReferenceCountedImpl<Jrd::JAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IVersionedImpl<Jrd::JAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IAttachment> > > > >::cloopexecuteDispatcher (
    self=0x7ffff507ebe8, status=0x7ffff13dd908, transaction=0x0, stmtLength=15, 
    sqlStmt=0x7ffff7700d10 "SET TRANSACTIONrg_tx_start activeinct_count_0 from tlogrt position 0\n    as\n    begin\n      execute statement ('insert into tlog(msg) values(?)') ('Tx start');\n    end\n    \314\314\314\314", '\335' <repeats 15 times>, "݀\366\205UUU", dialect=3, inMetadata=0x0, inBuffer=0x0, outMetadata=0x0, outBuffer=0x0)
    at /workspace/rdb/firebird_master/src/include/firebird/IdlFbInterfaces.h:10401
#6  0x00007ffff7bc24b0 in Firebird::IAttachment::execute<Firebird::CheckStatusWrapper> (this=0x7ffff507ebe8, status=0x7ffff13dd900, transaction=0x0, stmtLength=15, 
    sqlStmt=0x7ffff7700d10 "SET TRANSACTIONrg_tx_start activeinct_count_0 from tlogrt position 0\n    as\n    begin\n      execute statement ('insert into tlog(msg) values(?)') ('Tx start');\n    end\n    \314\314\314\314", '\335' <repeats 15 times>, "݀\366\205UUU", dialect=3, inMetadata=0x0, inBuffer=0x0, outMetadata=0x0, outBuffer=0x0)
    at /workspace/rdb/firebird_master/src/include/firebird/IdlFbInterfaces.h:2233
#7  0x00007ffff7baf26d in Why::YAttachment::execute (this=0x7ffff7a99850, status=0x7ffff13dd900, transaction=0x0, length=15, 
    string=0x7ffff7700d10 "SET TRANSACTIONrg_tx_start activeinct_count_0 from tlogrt position 0\n    as\n    begin\n      execute statement ('insert into tlog(msg) values(?)') ('Tx start');\n    end\n    \314\314\314\314", '\335' <repeats 15 times>, "݀\366\205UUU", dialect=3, inMetadata=0x0, inBuffer=0x0, outMetadata=0x0, outBuffer=0x0)
    at /workspace/rdb/firebird_master/src/yvalve/why.cpp:5695
#8  0x00007ffff7be937f in Firebird::IAttachmentBaseImpl<Why::YAttachment, Firebird::CheckStatusWrapper, Firebird::IReferenceCountedImpl<Why::YAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IVersionedImpl<Why::YAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IAttachment> > > > >::cloopexecuteDispatcher (
    self=0x7ffff7a99858, status=0x7ffff13dda98, transaction=0x0, stmtLength=15, 
    sqlStmt=0x7ffff7700d10 "SET TRANSACTIONrg_tx_start activeinct_count_0 from tlogrt position 0\n    as\n    begin\n      execute statement ('insert into tlog(msg) values(?)') ('Tx start');\n    end\n    \314\314\314\314", '\335' <repeats 15 times>, "݀\366\205UUU", dialect=3, inMetadata=0x0, inBuffer=0x0, outMetadata=0x0, outBuffer=0x0)
    at /workspace/rdb/firebird_master/src/include/firebird/IdlFbInterfaces.h:10401
#9  0x00005555555bc64c in Firebird::IAttachment::execute<Firebird::CheckStatusWrapper> (this=0x7ffff7a99858, status=0x7ffff13dda90, transaction=0x0, stmtLength=15, 
    sqlStmt=0x7ffff7700d10 "SET TRANSACTIONrg_tx_start activeinct_count_0 from tlogrt position 0\n    as\n    begin\n      execute statement ('insert into tlog(msg) values(?)') ('Tx start');\n    end\n    \314\314\314\314", '\335' <repeats 15 times>, "݀\366\205UUU", dialect=3, inMetadata=0x0, inBuffer=0x0, outMetadata=0x0, outBuffer=0x0)
    at /workspace/rdb/firebird_master/src/include/firebird/./IdlFbInterfaces.h:2233
#10 0x00005555555c75c9 in rem_port::execute_immediate (this=0x7ffff7185e50, op=op_exec_immediate, exnow=0x7ffff27fa790, sendL=0x7ffff27f9ee8)
    at /workspace/rdb/firebird_master/src/remote/server/server.cpp:3452
#11 0x00005555555cce80 in process_packet (port=0x7ffff7185e50, sendL=0x7ffff27f9ee8, receive=0x7ffff27fa488, result=0x7ffff13ddd48)
    at /workspace/rdb/firebird_master/src/remote/server/server.cpp:4935
#12 0x00005555555d23f6 in loopThread () at /workspace/rdb/firebird_master/src/remote/server/server.cpp:6481
#13 0x00005555555f5746 in (anonymous namespace)::ThreadArgs::run (this=0x7ffff13dde80) at /workspace/rdb/firebird_master/src/common/ThreadStart.cpp:78
#14 0x00005555555f57f9 in (anonymous namespace)::threadStart (arg=0x7ffff76fe400) at /workspace/rdb/firebird_master/src/common/ThreadStart.cpp:94
#15 0x00007ffff790b669 in start_thread (arg=<optimized out>) at pthread_create.c:479
#16 0x00007ffff7831323 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

```

hvlad commented:
> I've noticed test bugs.core_5533 hanging after merging this branch. In debug it crashing on assert. Git bisect tells it broken in 1c844b2. May be some regression surfaced after these changes?

Thanks ! Actually, you found a bug that show itself when:
- transaction is started using SET TRANSACTION SQL statement, i.e. not via API
- there is ON TRANSACTION START trigger
- trigger run EXECUTE STATEMENT against local transaction (that is currently starting).

I'm fixing it.
